specs/agents/tracing-api-otel.md

```diff
@@ -0,0 +1,277 @@
+
+For both transactions and spans, an optional `otel.span_kind` property will be provided by agents when set through
+the OTel API.
+This value should be stored into Elasticsearch documents to preserve OTel semantics and help future OTel integration.
+
+Possible values are `CLIENT`, `SERVER`, `PRODUCER`, `CONSUMER` and `INTERNAL`, refer to [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind) for details on semantics.
+
+By default, OTel spans have their `SpanKind` set to `INTERNAL` by OTel API implementation, so it is assumed to always be provided when using the bridge.
+
+For existing agents without OTel bridge or for data captured without the bridge, the APM server has to infer the value of `otel.span_kind` with the  following algorithm:
+
+```javascript
+span_kind = null;
+if (isTransaction(item)) {
+    if (item.type == "messaging") {
+        span_kind = "CONSUMER";
+    } else if (item.type == "request") {
+        span_kind = "SERVER";
+    }
+} else {
+    // span
+    if (item.type == "external" || item.type == "storage" || item.type == "db") {
+        span_kind = "CLIENT";
+    }
+}
+
+if (span_kind == null) {
+    span_kind = "INTERNAL";
+}
+
+```
+
+While being optional, inferring the value of `otel.span_kind` helps to keep the data model closer to the OTel specification, even if the original data was sent using the native agent protocol.
+
+### Span status
+
+OTel spans have a [Status](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status)
+field to indicate the status of the underlying task they represent.
+
+When the [Set Status](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status) on OTel API is used, we can map it directly to `span.outcome`:
+- OK => Success
+- Error => Failure
+- Unset (default) => Unknown
+
+However, when not provided explicitly agents can infer the outcome from the presence of a reported error.
+This behavior is not expected with OTel API with status, thus bridged spans/transactions should NOT have their outcome
+altered by reporting (or lack of reporting) of an error. Here the behavior should be identical to when the end-user provides
+the outcome explicitly and thus have higher priority over the inferred value.
+
+### Attributes mapping
+
+OTel relies on key-value pairs for span attributes.
+Keys and values are protocol-specific and are defined in [semantic convention](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions) specification.
+
+In order to minimize the mapping complexity in agents, most of the mapping between OTel attributes and agent protocol will be delegated to APM server:
```

Comment:
Makes sense, but looking at the inference algorithm (which is an additional one to one we already have), and result/outcome inference - it seems reasonable to reconsider moving more login from agents to the server.
Not blocking this PR of course, but I think something to reconsider

Reply:
I agree with you, being able to delegate all of the mapping and only implement it once would be great.

The inference algorithm on the agent is only required for the "essential" fields as they are reused by other agent features.
- type (and subtype) are required for breakdown metrics: making this feature otel-compliant seems challenging.
- outcome is required as we map it from otel status and don't store it as-is. (in hindsight maybe copying this one to `otel.status` next to `otel.attributes` would have been a good idea.